[SPARK-9312] [ML] Added max confidence factor to OneVsRestModel

badriub opened the pull request:
1. I've declared a new method  predictPointWithProbability in GeneralizedLinearModel and implemented the method only for LogisticRegressionModel. 
2. I calculate the probability using the odds ratio.
3. For the other implementations of GeneralizedLinearModel, I have added an "not implemented exception". I can look into how to calculate probability if needed.

WRT OVR,
I've extracted the rawProbability as the maxProbability of all classes and added a new column in the returned dataframe.
I've fixed the test to include the new column "rawProbability"

badriub committed:
Added predictive probability to OneVsRestModel and LogisticRegressionModel

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/regression/Lasso.scala:
```diff
@@ -37,6 +40,13 @@ class LassoModel (
+                                       dataMatrix: Vector,
+                                       weightMatrix: Vector,
+                                       intercept: Double) = {
+    throw new Exception("Not implemented for LassoModel")
```
Without commenting on the suitability of the PR -- I can at least say that `throw new Exception` is never a great idea. Use `UnsupportedOperationException` here, for example.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/regression/Lasso.scala:
```diff
@@ -17,7 +17,10 @@
 
 package org.apache.spark.mllib.regression
 
+import java.lang
```
Never needs to be imported.

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/classification/SVM.scala:
```diff
@@ -73,6 +73,13 @@ class SVMModel (
   }
 
+  override protected def predictPointWithProbability(
+                                                      dataMatrix: Vector,
```
Indentation is wrong here and elsewhere

srowen commented on mllib/src/main/scala/org/apache/spark/mllib/classification/LogisticRegression.scala:
```diff
@@ -114,6 +114,51 @@ class LogisticRegressionModel (
+        case None => score
+      }, score)
+    } else {
+      /**
```
Nit: this isn't javadoc

srowen commented on mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala:
```diff
@@ -19,6 +19,8 @@ package org.apache.spark.ml.classification
 
 import java.util.UUID
 
+import org.apache.spark.ml.param.shared.HasRawPredictionCol
```
Nit: this import should go with spark imports

jkbradley commented:
@badriub I just added some thoughts on the JIRA.  Could you please take a look and respond there?  Thanks!

badriub commented:
I have replied on the JIRA, please take a look. Thanks.

badriub committed:
SPARK-9312: Adding confidence factor to OneVsRest Model

badriub committed:
SPARK-9312: Undoing changes for the LogisticRegression related to
predictive probability since it was not needed.

badriub committed:
Implementing review comments

badriub committed:
Correcting indentation

badriub commented on mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala:
```diff
@@ -19,23 +19,24 @@ package org.apache.spark.ml.classification
 import org.apache.spark.storage.StorageLevel
 
+import scala.language.existentials
+
```
Used the Intellj importer plugin as recommended on the style guide.

jkbradley replied:
You may need to adjust the import organizer settings, though.  The order should be Java, Scala, other, Spark.

badriub commented on mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala:
```diff
@@ -127,10 +128,16 @@ final class OneVsRestModel private[ml] (
-      .drop(accColName)
+      .withColumn($(rawPredictionCol), probabilityUDF(col(accColName)).as($(rawPredictionCol),
+      labelMetadata)).drop(accColName)
   }
```
Adding the max confidence factor as output column.

badriub commented:
@jkbradley : I've reverted the changes for LogisticRegression (for entire class hierarchy) since they were not needed. I've updated JIRA and PR as discussed.
@srowen : I've used the importer plugin and made sure I follow the style guide.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala:
```diff
@@ -19,23 +19,24 @@ package org.apache.spark.ml.classification
  * Params for [[OneVsRest]].
  */
-private[ml] trait OneVsRestParams extends PredictorParams {
+private[ml] trait OneVsRestParams extends PredictorParams with HasRawPredictionCol {
```
It will be easier for you if you have OneVsRestParams extend ClassifierParams.  That will include the rawPredictionCol, plus the updated validateAndTransformSchema method.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala:
```diff
@@ -127,10 +128,16 @@ final class OneVsRestModel private[ml] (
       predictions.maxBy(_._2)._1.toDouble
     }
 
+    // output the index of the classifier with highest confidence as prediction
```
fix doc

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala:
```diff
@@ -127,10 +128,16 @@ final class OneVsRestModel private[ml] (
     }
 
+    // output the index of the classifier with highest confidence as prediction
+    val probabilityUDF = udf { (predictions: Map[Int, Double]) =>
```
This is outputting a Double, but it should return a Vector with one value for each class label.  Actually, the output value should be the accCol converted to a Vector.

badriub replied:
Sorry for the delayed response. I've made some small changes as per your recommendations. WRT this point, I'm not sure I follow. As far as I understand, I need to return the highest confidence factor(which will be Double) corresponding to the model used with OVR.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala:
```diff
@@ -220,4 +227,4 @@ final class OneVsRest(override val uid: String)
     copied
   }
-}
+}
```
missing newline

jkbradley commented:
That should do it for a first pass.  I'll check back for updates.  Thanks!
Also, can you please update:
* the PR description (the first comment in this PR) to match the more recent changes, and
* the title to use the tag "ml" instead of "mllib"?

badriub committed:
Merge branch 'master' of https://github.com/badriub/spark.git

badriub committed:
Import order corrected

badriub committed:
Missed fixing a doc earlier

viktortnk commented:
@jkbradley @badriub Any updates on this issue?

From diff I see that `rawPrediction` has `Double` type. Should it return an Array of values from each binary classifier

I'm happy to help if required to get it through

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AxenGitHub commented:
Is there any news on this branch? we would benefit a lot from this feature.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

jkbradley commented:
Hi, sorry I let this PR get stale.  This should be resolved now by https://github.com/apache/spark/pull/21044 so would you mind closing this issue @badriub ?  Thanks though!

AmplabJenkins commented:
Can one of the admins verify this patch?
